### ISSUE TITLE
3.1-rc1 libsepol: still something is wrong when LTO is used

### ISSUE THREAD

**Issue by kloczek** [SELinuxProject/selinux#245]
Here is command which shows which one dynamic symbols has been chopped when LTO is used:
```
[tkloczko@barrel SPECS]$ diff -u <(nm -D libsepol.so.1 | grep -w T|grep -w T|awk '{print $3}'|sort) <(nm -D /usr/lib64/libsepol.so.1|grep -w T|awk '{print $3}'|sort)
--- /dev/fd/63  2020-05-19 04:17:44.930203098 +0100
+++ /dev/fd/62  2020-05-19 04:17:44.939203118 +0100
@@ -1,14 +1,6 @@
 cil_add_file
-cil_build_policydb
-cil_build_policydb
-cil_compile
-cil_compile
 cil_db_destroy
 cil_db_init
-cil_filecons_to_string
-cil_filecons_to_string
-cil_selinuxusers_to_string
-cil_selinuxusers_to_string
 cil_set_attrs_expand_generated
 cil_set_attrs_expand_size
 cil_set_disable_dontaudit
@@ -21,8 +13,6 @@
 cil_set_policy_version
 cil_set_preserve_tunables
 cil_set_target_platform
-cil_userprefixes_to_string
-cil_userprefixes_to_string
 cil_write_policy_conf
 expand_module_avrules
 sepol_bool_clone
```
Looks like all those chopped symbols have boon listed in sym file two times:
```
[tkloczko@barrel src]$ grep cil_build_policydb libsepol.map
        cil_build_policydb;
        cil_build_policydb;

[tkloczko@barrel src]$ grep cil_db_init libsepol.map
        cil_db_init;
```
I'm not sure is it wrong definition in sym file or it is linker bug ..

**Comment by fishilico** [comment]
Hi,
These symbols are defined twice in https://github.com/SELinuxProject/selinux/blob/20200518/libsepol/src/libsepol.map.in because their API/ABI changed and libsepol uses symbol versionning. For example with `cil_build_policydb`:
```
LIBSEPOL_1.0 {
  global:
	cil_build_policydb;
...
}
LIBSEPOL_1.1 {
  global:
	cil_build_policydb;
...
}
```
This symbol is defined using `.symver` macros in https://github.com/SELinuxProject/selinux/blob/20200518/libsepol/cil/src/cil.c#L59-L60 :
```C
asm(".symver cil_build_policydb_pdb,        cil_build_policydb@LIBSEPOL_1.0");
asm(".symver cil_build_policydb_create_pdb, cil_build_policydb@@LIBSEPOL_1.1");
```

As I already told you 6 months ago in https://github.com/SELinuxProject/selinux/issues/165#issuecomment-533811844, 
> it is a known issue that compiling with LTO to drops some symbols, as LTO does not support symbol versioning properly.

Since September there seems to have been some progress in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=48200#c28, with a gcc patch in December, https://gcc.gnu.org/legacy-ml/gcc-patches/2019-12/msg01162.html. But this seems to be still a work in progress in GCC.

In short: "not our bug, LTO is known to be broken, people are trying to fix it and it takes time".

**Comment by kloczek** [comment]
In that sym file fragment which you've quoted you fiorgot about one cruccial fragment:
```
LIBSEPOL_1.0 {
  global:
	cil_build_policydb;
...
}
LIBSEPOL_1.1 {
  global:
	cil_build_policydb;
...
}
```
That fragment actually looks different:
```
LIBSEPOL_1.0 {
  global:
	cil_build_policydb;
...
}
LIBSEPOL_1.1 {
  global:
	cil_build_policydb;
...
} LIBSEPOL_1.0;
```
Difference is in last line which is causing that all `LIBSEPOL_1.1` symbols inherits all `LIBSEPOL_1.0` .. In that case listing exactly the same symbols two times looks like is causing some confusion (this is why I mention that it may be some ld issue).

**Comment by kloczek** [comment]
So looking on code I don't see any reasons to for example cil_build_policydb() which indicates that tgis routine is working differently because it was part of the 1.1 and `1.0 ABI. I've removed those duplicates from 1.1 interface description because I think that those symbols because they are part of the 1.0 ABI.

Nevertheless after removing those duplicates and have them only in  1.0 section seems solved issue.
That is actual patch which I've used:
```
--- a/libsepol/src/libsepol.map.in~     2020-05-15 14:54:08.000000000 +0100
+++ b/libsepol/src/libsepol.map.in      2020-05-19 04:25:48.335295650 +0100
@@ -254,11 +254,6 @@

 LIBSEPOL_1.1 {
   global:
-       cil_build_policydb;
-       cil_compile;
-       cil_userprefixes_to_string;
-       cil_selinuxusers_to_string;
-       cil_filecons_to_string;
        cil_set_target_platform;
        cil_set_policy_version;
        cil_set_mls;
```
Feel free to use this.
So after fixing libsepol 3.1-rc1 I'm still working with old binaries of of libsemanage 3.0 and libselinux and going now to have look closer on libselinux 3.1-rc-1 (then libsemanage 3.1.-rc1).

Nevertheless libsepol with LTO is ~50% smaller so it it is a lot of bits which could be removed from libsepol without any consequences.

To asses what could be removed from the actiuall code I think that linker diagnostics switched on -Wl,--debug=2 should show what exact actions linger performed to reduce so much actual size DSO produced with LTO.

**Comment by fishilico** [comment]
> So looking on code I don't see any reasons to for example cil_build_policydb() which indicates that tgis routine is working differently because it was part of the 1.1 and `1.0 ABI.

The code defines aliases that depend on the version (https://github.com/SELinuxProject/selinux/blob/20200518/libsepol/cil/src/cil.c#L59-L60): * `cil_build_policydb@LIBSEPOL_1.0` is function `cil_build_policydb_pdb()`, that calls `cil_binary_create_allocated_pdb()`
* `cil_build_policydb@@LIBSEPOL_1.1` is function `cil_build_policydb_create_pdb()`, that calls `cil_binary_create()`.

The difference between these functions is explained in a comment, https://github.com/SELinuxProject/selinux/blob/20200518/libsepol/cil/src/cil_binary.c#L4794-L4795 :
```C 
// assumes policydb is already allocated and initialized properly with things
// like policy type set to kernel and version set appropriately
int cil_binary_create_allocated_pdb(const struct cil_db *db, sepol_policydb_t *policydb)
```

About the fact that I missed that the version script file contains
```
LIBSEPOL_1.1 {
  global:
	cil_build_policydb;
...
} LIBSEPOL_1.0;
```
this is actually what is documented in `ld`'s official documentation: https://sourceware.org/binutils/docs-2.34/ld/VERSION.html. The `LIBSEPOL_1.0;` here seems to be used to indicate some sort of dependency relationship and may be important for the linker to use the right order of the blocks.
Anyway, in the documentation there does not seem to be duplicated symbols (only new ones in each version block) in the example version script. The documentation also gives an example of a symbol `foo` which has several versions (and presents `__asm__(".symver ")` macros). So `libsepol` could be doing something wrong by duplicating names in `libsepol.map.in`, which could be fixed with your last patch.

In order to test it, can you please run `objdump -T libsepol/src/libsepol.so |grep -e cil_build_policydb -e cil_compile -e cil_filecons_to_string -e cil_selinuxusers_to_string -e cil_userprefixes_to_string` (on the produced `libsepol.so` file) and check that the output contains symbols with two versions each, with lines such as these ones?
```
0000000000051e30 g    DF .text	000000000000000b (LIBSEPOL_1.0) cil_selinuxusers_to_string
0000000000051c70 g    DF .text	00000000000001bf  LIBSEPOL_1.1 cil_selinuxusers_to_string
0000000000051c60 g    DF .text	000000000000000b (LIBSEPOL_1.0) cil_userprefixes_to_string
0000000000051b60 g    DF .text	00000000000000f1  LIBSEPOL_1.1 cil_userprefixes_to_string
00000000000510a0 g    DF .text	0000000000000163  LIBSEPOL_1.1 cil_compile
0000000000051210 g    DF .text	0000000000000005 (LIBSEPOL_1.0) cil_compile
0000000000051220 g    DF .text	000000000000005e (LIBSEPOL_1.0) cil_build_policydb
0000000000051280 g    DF .text	000000000000005e  LIBSEPOL_1.1 cil_build_policydb
0000000000052230 g    DF .text	000000000000000b (LIBSEPOL_1.0) cil_filecons_to_string
0000000000051e40 g    DF .text	00000000000003ed  LIBSEPOL_1.1 cil_filecons_to_string
```

**Comment by kloczek** [comment]
So IMO it looks like new  LTO issue :/

Are you sure that you want to do that that want? IMO it is oversimplification everything and provide some backward compatibility which could be easily solved by SONAME bulm.
SELinix libraries are not so critical like libc so IMO it only puffs binary code and is king asking for troubles on maintaining internal bigger nephropathy .. especially if you wiill add to that picture that libsepol is only used by SELInux binaries.
```
[root@barrel RPMS]# dnf repoquery --whatrequires 'libsepol.so.1()(64bit)'
Last metadata expiration check: 0:10:42 ago on Wed 20 May 2020 18:15:05 BST.
libselinux-utils-0:3.0-4.fc33.x86_64
libsemanage-0:3.0-3.fc32.x86_64
libsepol-devel-0:3.0-3.fc32.x86_64
parted-0:3.3-3.fc33.x86_64
policycoreutils-0:3.0-2.fc32.x86_64
python3-setools-0:4.3.0-1.fc33.x86_64
secilc-0:3.0-2.fc32.x86_64
```
From that list even parted can be dropped because it does not uses libsepol at all (it is only linked with libsepol but none of the sympols from that library is used).

In other words necessity to have that kind of duplicated symbols DSO after upgrade all SELinux binaries will be NULL!!

IMO all that code can be removed especially looking on version that it was on 1.0 to 1.1 transition so it was sooooo long time ago that today probably number of systems with that old binaries will NONE.

Do you that now?

More ... looking on fact that only few binaries are using libsepol I can guess that many symbols from that library quite possible that could be used only one time so keeping some routines as parto of the library quite possible does not make any sense.

You have way better working knowledge what is used in SELinux by what so you can probably quite precisely asses am I could be right or not :)

However IMO it would be good to strictly follow KISS principle on maintaining SELinux user space tooling :P

**Comment by fishilico** [comment]
I did not understand your reaction. Is LTO broken with symbolic versioning as used by `libsepol` or not?

Anyway, if you want to disable the symbolic versioning feature, you can build `libsepol.so` with `-DDISABLE_SYMVER=1` in `CFLAGS`, because the whole problematic code is in a `#ifndef DISABLE_SYMVER` block: https://github.com/SELinuxProject/selinux/blob/20200518/libsepol/cil/src/cil.c#L58

**Comment by bachradsusi** [comment]
FYI https://lore.kernel.org/selinux/20200930145031.910190-1-plautrba@redhat.com/T/#t should fix this issue
